### PR TITLE
refactor: protocol clean up and extension for dataFormat

### LINKED ISSUE #72: Refactor and extend the DWN protocol for supporting different VC format
# Feature Request

## Which Areas Would Be Affected?
_e.g., build, extension, etc._

## Why Is the Feature Desired?
_Are there any requirements?_

## Solution Proposal
_If possible, provide a (brief!) solution proposal._

## Type of Issue
_i.e., new feature, improvement, cleanup, etc._

## Checklist

- [ ] assigned appropriate label?
- [x] **Do NOT select a milestone or an assignee!**

### PR DESCRIPTION
**Pull request by wolf4ood**
## What this PR changes/adds

Refactors a bit the protocol by removing some unused fields and added support
for plugging and validating data base on it's format

## Why it does that

Clean up and extensibility

## Further notes

Those fields where removed in the protocol since were not part of the spec

- MessageResponseObject : `messageId`
- RequestObject: `target` and `requestId`
- Descriptor: `nonce`

Previously the data was assumed to be a Verifiable credential in JWT format.
Now when  sending messages now the property in the Descriptor `dataFormat` is required and it will drive
the data validation introduced with `DataValidator`.

This `dataFormat` is also stored along with the data itself in the IHStore and a new object `Record` is introduced and returned when issuing a `CollectionsQuery` request. 
By doing so the client knows which kind of data is returned from the request and act accordingly.

## Linked Issue(s)

Closes #72 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

**Review by paullatzelsperger — APPROVED** [review]

File: core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java [hunk 1]
```diff
@@ -133,16 +138,42 @@ public StatusResult<Void> addVerifiableCredential(String hubBaseUrl, SignedJWT v
             if (response.code() != 200) {
                 return identityHubCallError(response);
             }
+            var responseObject = objectMapper.readValue(response.body().byteStream(), ResponseObject.class);
+
+            if (responseObject.getStatus() != null && !responseObject.getStatus().isSuccess()) {
+                return StatusResult.failure(ResponseStatus.FATAL_ERROR, responseObject.getStatus().getDetail());
+            }
+
+            var requestStatus = Optional.of(responseObject.getStatus())
```
Comment (bscholtes1A) [inline core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java:147]: try with resources?

File: core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java [hunk 2]
```diff
@@ -133,16 +138,42 @@ public StatusResult<Void> addVerifiableCredential(String hubBaseUrl, SignedJWT v
...
+                    .map(status -> Stream.of(status))
+                    .orElseGet(Stream::empty);
+
+            var messageStatus = responseObject.getReplies()
```
Comment (bscholtes1A) [inline core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java:151]: try with resources?

File: core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java [hunk 3]
```diff
@@ -133,16 +138,42 @@ public StatusResult<Void> addVerifiableCredential(String hubBaseUrl, SignedJWT v
...
+                    .stream()
+                    .map(MessageResponseObject::getStatus);
+
+            return Stream.concat(messageStatus, requestStatus)
```
Comment (bscholtes1A) [inline core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java:155]: The logic in this new block of code lost me... :)
Perhaps would be better to encapsulate into a dedicated method documenting the behavior.

File: core/identity-hub/src/main/java/org/eclipse/edc/identityhub/processor/data/JwtVerifiableCredential.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import org.eclipse.edc.identityhub.spi.credentials.model.VerifiableCredential;
+import org.eclipse.edc.identityhub.spi.processor.data.DataValidator;
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.result.Result;
+
+import java.util.Optional;
+
+public class JwtVerifiableCredential implements DataValidator {
+
+    public static final String DATA_FORMAT = "application/vc+jwt";
```
Comment (bscholtes1A) [inline core/identity-hub/src/main/java/org/eclipse/edc/identityhub/processor/data/JwtVerifiableCredential.java:28]: I know it is not straightforward due to the client/identityhub split in this repo, but could we consider centralizing these VC data formats into a dedicated file in the `spi` folder?

File: spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/processor/data/DataValidator.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+     * @param data Input
+     * @return The result of the validation
+     */
+    Result<Void> validate(byte[] data);
+
+
+    /**
+     * Returns the Media type that implementor of {@link DataValidator} is able to validate.
+     */
+    String dataFormat();
```
[the remark this answers is not in the document]
Comment (wolf4ood) [reply spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/processor/data/DataValidator.java:39]: It's used when registering the Validator in the Registry

https://github.com/eclipse-edc/IdentityHub/pull/74/files#diff-78d16f3063ae987cc140351b184e47c53695f3f6e08f3db1025c1000fdd0d5a5R34-R36

File: core/identity-hub/src/main/java/org/eclipse/edc/identityhub/processor/data/JwtVerifiableCredential.java (hunk 4 above)
Comment (wolf4ood) [reply core/identity-hub/src/main/java/org/eclipse/edc/identityhub/processor/data/JwtVerifiableCredential.java:28]: Yeah we can definitely, i would do it in a separated PR though.

File: core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java (hunk 3 above)
Comment (wolf4ood) [reply core/identity-hub-client/src/main/java/org/eclipse/edc/identityhub/client/IdentityHubClientImpl.java:155]: Yeah i can move the result handling in a separated method with doc.

**Review by bscholtes1A — APPROVED** [review]

**Pull request merged**